### DISCUSSION TITLE
Clarification requested regarding DMI: Random object created and used only once

### DISCUSSION THREAD

**Discussion by soddencarpenter** [spotbugs/spotbugs#1628] [Q&A]
Hi,

We have some code somewhat similar to this simple case:
```
public class ExampleRndClass
{
  private static final Random RND = new SecureRandom();
  private static final ExampleRndClass INST = new ExampleRndClass();
  
  private ExampleRndClass() {}
 
  public ExampleRndClass getInstance()
  {
    return INST;
  }
    
  public String getStr()
  {
    return String.format("Hi There-%d",
        RND.nextInt());
  }
}
```

SpotBugs is marking the line `RND.nextInt()` with

> Bug: Random object created and used only once in ExampleRndClass.getStr()
>
> This code creates a java.util.Random object, uses it to generate one random number, and then discards the Random object. This produces mediocre quality random numbers and is inefficient. If possible, rewrite the code so that the Random object is created once and saved, and each time a new random number is required invoke a method on the existing Random object to obtain it.  
>If it is important that the generated Random numbers not be guessable, you 
must not create a new Random for each random number; the values are too easily guessable. You should strongly consider using a java.security.SecureRandom instead (and avoid allocating a new SecureRandom for each random number needed).  
> 
>Rank: Troubling (14), 
>confidence: High
>
>Pattern: DMI_RANDOM_USED_ONLY_ONCE 
>
>Type: DMI, 
>Category: BAD_PRACTICE (Bad practice)<hr size="1" />

The Random object is not created in the method; it is a `static final` for the class. 

Consequently, it is unclear what the error is indicating, and what the mitigation should be.

This appears to have been flagged with 4.3.0.

What should be the approach here? Thanks!

(note: migrated from the https://github.com/spotbugs/discuss/issues/116; though there is an pinned announcement that things will be moved, it wasn't clear that one should be using this location, and I didn't see anything in the announcement actually directing one to this discussion area).

**Comment by mkienenb** [comment]
Does the exact code you posted generate that error?  I'm not seeing a warning when I run it through spotbugs.

If not, then there must be something different with your code than with the code posted.

My guess would be that the method using RND has been determined to only be callable once.

**Reply by soddencarpenter** [reply]
Hi,

Thanks for the reply.

Yes, that exact code above is generating the warning at the `RND.nextInt();`. I can see the error marker in the Eclipse version of Spotbugs, but we got into this question because our original code was flagged by our CI system for using essentially the same approach. I was also able to reproduce the issue in the GUI using the code above, as shown in the image.

I am happy to admit we might be doing something incorrectly, but it just is not clear to me what that would be. The `Random` is not created in the method, and I do not see how the analyzer can make any determination about the number of external calls being applied to it. In our actual code case, the method is called hundreds of times.

<img width="1200" alt="2021-08-01 12_31_20-SpotBugs - Blah" src="https://user-images.githubusercontent.com/15717543/127781732-7df51691-cf54-4d1f-82ff-7a4507de8afc.png">

I am also curious about your different outcome.

**Reply by mkienenb** [reply]
The code you provided looks fine to me.  Having a static instance and a public method using it seems like it would be good enough.

I dropped the class you provided into an existing project using spotbugs.  Then I used spotbugs-4.0.1-SNAPSHOT plus findbugs-sec-plugin and there were a bunch of excludes defined, which could account for the difference in output.   It's what I have handy at the moment.

So I just ran it a second time from the command line using just your class an 4.0.1 snapshot and also got no warnings.

java -jar $SPOTBUGS_HOME/lib/spotbugs.jar -textui -low -onlyAnalyze spotbugs.ExampleRndClass ../WEB-INF/classes/spotbugs/ExampleRndClass.class 

I also put some obvious errors in the file to verify that it really did output warnings if something was wrong, which it did.

Perhaps there's no check for Random in the version I'm currently using.

So I downloaded 4.3.0, and tried it.  Now I do get warning:

H B DMI: Random object created and used only once in spotbugs.ExampleRndClass.getStr()  At ExampleRndClass.java:[line 27]

And there's a few issues for this (apparently started happening in 4.2.2.), but is reported to be fixed in https://github.com/spotbugs/spotbugs/pull/1540.

Since 4.3.0 was released July 1st, I'd say that the problem still isn't fixed.  You could try it after building from source just to be sure it wasn't fixed last month.

I'm guessing #1540 is probably a good place to start looking if you want to submit a patch for it yourself, or even just to create a testcase demonstrating the issue.

**Reply by soddencarpenter** [reply]
Thank you so much for the extended effort. I did not consider that the issue could be anything but the way we were doing something. I will try a trunk compile, and if it isn't resolved I will follow up on that #1540 ticket.

Thanks again!

**Comment by hazendaz** [comment]
seems resolved, closing.
